File: DittoDatto-old/packages/mercury-engine/src/core/bookings/booking.ts
// packages/functions/src/MercuryEngine/booking.ts
// Core Domain: Booking Context - Create Booking from Hold
// Migrated from: .docs/project-context/BookingScripts/src/services/booking/create.ts

import { getAuth } from "firebase-admin/auth";
import { db } from "../../config/firebase.js";
import { Booking, Hold, Service, BookingPolicySchema } from "@dittodatto/shared-types";
import { upsertCustomerFromBooking } from "../shared/customer.js";

import { HttpsError } from "../shared/errors.js";

// ============================================================================
// Pure Functions (testable without Firestore)
// ============================================================================

/**
 * Input data for building a booking receipt.
 * All fields are plain data — no Firestore refs.
 */
export interface BookingReceiptInput {
  hold: Hold;
  services: (Service & { id: string })[];
  userName: string;
  userEmail: string;
  storeName: string;
  paymentId: string;
  notes?: string;
}

/**
 * Builds a Booking object (the receipt/snapshot) from hold + services + user data.
 * Pure function — no IO. All data passed in, Booking object returned.
 *
 * Uses the Snapshot Pattern: copies service price/title at booking time
 * for historical accuracy (prices can change later).
 */
export function buildBookingReceipt(input: BookingReceiptInput): Booking & { storeName: string } {
  const { hold, services, userName, userEmail, paymentId, storeName, notes } = input;

  const holdServiceIds: string[] = hold.serviceIds
    || (hold.serviceId ? [hold.serviceId] : []);

  // Aggregate totals from all services
  const totalDuration = services.reduce((sum, s) => sum + (s.duration || 60), 0);
  const totalPrice = services.reduce((sum, s) => sum + (s.price || 0), 0);
  const serviceTitle = services.map((s) => s.title).join(", ");

  // Calculate Start/End Time as timezone-naive local times.
  const startTime = `${hold.date}T${hold.slotTime}:00`;

  const [startH, startM] = hold.slotTime.split(":").map(Number);
  const endTotalMinutes = startH * 60 + startM + totalDuration;
  const endH = String(Math.floor(endTotalMinutes / 60) % 24).padStart(2, "0");
  const endM = String(endTotalMinutes % 60).padStart(2, "0");
  const endTime = `${hold.date}T${endH}:${endM}:00`;

  // Build items array — one entry per service
  const items = services.map((s) => ({
    serviceId: s.id,
    title: s.title,
    price: s.price,
    duration: s.duration,
    ...(hold.staffId ? { staffId: hold.staffId } : {}),
    storeId: hold.storeId,
    companyId: hold.companyId,
  }));

  const booking: Booking = {
    id: paymentId,
    status: "confirmed",
    userId: hold.userId,
    companyId: hold.companyId,
    storeId: hold.storeId,
    serviceId: holdServiceIds[0],
    ...(hold.staffId ? { staffId: hold.staffId } : {}),
    ...((hold as any).resourceId ? { resourceId: (hold as any).resourceId } : {}),
    serviceTitle,
    duration: totalDuration,
    priceAtTimeOfBooking: totalPrice,
    currency: services[0].currency,
    userName,
    userEmail,
    startTime: startTime as unknown as Date,
    endTime: endTime as unknown as Date,
    createdAt: new Date() as unknown as Date,
    items,
    addonResourceIds: [],
    paymentId,
    channel: "app",
    attendeeCount: 1,
    ...(notes?.trim() ? { notes: notes.trim() } : {}),
  };

  return { ...booking, storeName };
}

/**
 * Input for cancellation policy check.
 */
export interface CancellationCheckInput {
  bookingStartTime: string | Date;
  policy: {
    clientCancelEnabled: boolean;
    minCancelNoticeHours: number;
  };
  now?: Date;
}

/**
 * Validates that a customer cancellation is allowed per the store's policy.
 * Pure function — throws HttpsError if policy forbids cancellation.
 *
 * Only called for customer-initiated cancellations (staff/admin bypass).
 */
export function checkCancellationPolicy(input: CancellationCheckInput): void {
  const { policy, now = new Date() } = input;

  if (!policy.clientCancelEnabled) {
    throw new HttpsError(
      "failed-precondition",
      "This business does not allow customer cancellations. Please contact the business directly.",
    );
  }

  if (policy.minCancelNoticeHours > 0) {
    const startTime = typeof input.bookingStartTime === 'string'
      ? new Date(input.bookingStartTime)
      : (input.bookingStartTime as any)?.toDate?.()
        ?? new Date(input.bookingStartTime as any);

    const deadlineMs = startTime.getTime() - (policy.minCancelNoticeHours * 60 * 60 * 1000);
    const deadline = new Date(deadlineMs);

    if (now.getTime() > deadlineMs) {
      throw new HttpsError(
        "failed-precondition",
        `Cancellation deadline has passed. Must cancel at least ${policy.minCancelNoticeHours} hours before the appointment (deadline: ${deadline.toISOString()}).`,
      );
    }
  }
}

/**
 * MercuryEngine - Create Booking from Hold
 * Converts a valid hold into a permanent booking using the Snapshot Pattern.
 * Copies service price/title at booking time for historical accuracy.
 *
 * @param holdId - The composite hold ID to convert
 * @param paymentId - Payment reference ID (used as booking ID for idempotency)
 * @returns Object with success status and bookingId
 */
export const createBookingFromHold = async (
  holdId: string,
  paymentId: string,
  notes?: string
) => {
  const holdRef = db.collection("holds").doc(holdId);

  // CRM data captured inside the transaction, used after it completes
  let crmData: {
    companyId: string; storeId: string; userId: string;
    userName: string; userEmail: string; userPhone?: string;
    staffId?: string; totalPrice: number; startTime: string;
    channel: string;
  } | null = null;

  // Atomic Transaction: Read Hold -> Create Booking -> Delete Hold
  await db.runTransaction(async (transaction) => {
    // A. Read the Hold
    const holdDoc = await transaction.get(holdRef);
    if (!holdDoc.exists) {
      throw new HttpsError("not-found", "Hold expired or not found");
    }
    const hold = holdDoc.data() as Hold;

    // B. Fetch Service Snapshots for the receipt
    const holdServiceIds: string[] = hold.serviceIds
      || (hold.serviceId ? [hold.serviceId] : []);

    if (holdServiceIds.length === 0) {
      throw new HttpsError("failed-precondition", "Hold has no service IDs");
    }

    const serviceSnaps = await Promise.all(
      holdServiceIds.map((sid: string) =>
        transaction.get(
          db.doc(
            `companies/${hold.companyId}/stores/${hold.storeId}/services/${sid}`
          )
        )
      )
    );

    const services = serviceSnaps.map((s) => ({
      ...(s.data() as Service),
      id: s.id,
    }));

    // Resolve user name/email
    let userName = "Kunde";
    let userEmail = "";
    try {
      const profileSnap = await transaction.get(db.doc(`users/${hold.userId}`));
      if (profileSnap.exists) {
        const profile = profileSnap.data();
        userName = profile?.name || "";
        userEmail = profile?.email || "";
      }
      if (!userName || userName === "Kunde") {
        const userRecord = await getAuth().getUser(hold.userId);
        userName = userName || userRecord.displayName || userRecord.email || "Kunde";
        userEmail = userEmail || userRecord.email || "";
      }
    } catch {
      // User might be deleted — fallback
    }

    // Fetch store name
    let storeName = "";
    try {
      const storeSnap = await transaction.get(
        db.doc(`companies/${hold.companyId}/stores/${hold.storeId}`)
      );
      if (storeSnap.exists) {
        storeName = storeSnap.data()?.name || "";
      }
    } catch {
      // Non-critical
    }

    // C. Build the receipt (pure)
    const bookingWithMeta = buildBookingReceipt({
      hold, services, userName, userEmail, storeName, paymentId, notes,
    });
    const bookingId = paymentId;

    // D. Write the Booking & Delete Hold (Atomic)
    const bookingRef = db.collection("bookings").doc(bookingId);
    transaction.set(bookingRef, bookingWithMeta);
    transaction.delete(holdRef);

    // E. Capture data for CRM upsert
    const totalPrice = bookingWithMeta.priceAtTimeOfBooking;
    const startTime = bookingWithMeta.startTime;
    crmData = {
      companyId: hold.companyId,
      storeId: hold.storeId,
      userId: hold.userId,
      userName,
      userEmail,
      staffId: hold.staffId,
      totalPrice,
      startTime: startTime as unknown as string,
      channel: "app",
    };
  });

  // CRM Side-Effect: Upsert customer record (fire-and-forget)
  // Runs AFTER the transaction so it never blocks or rolls back the booking.
  if (crmData) {
    const d: any = crmData;
    upsertCustomerFromBooking({
      companyId: d.companyId,
      storeId: d.storeId,
      userId: d.userId,
      userName: d.userName,
      userEmail: d.userEmail,
      bookingId: paymentId,
      staffId: d.staffId,
      priceAtTimeOfBooking: d.totalPrice,
      bookingTime: d.startTime,
      channel: d.channel,
    }).catch(() => {}); // Silently swallow — logged inside upsertCustomerFromBooking
  }

  return { success: true, bookingId: paymentId };
};

/**
 * MercuryEngine - Cancel Booking
 * Cancels an existing booking.
 * Bypasses client-side Firestore rules by operating via the server SDK.
 * 
 * @param bookingId - The ID of the booking to cancel
 * @param userId - The ID of the user requesting the cancellation (used for authorization)
 * @param reason - Optional reason for cancellation
 * @returns Object with success status
 */
export const cancelBooking = async (
  bookingId: string,
  userId: string,
  reason?: string
) => {
  const bookingRef = db.collection("bookings").doc(bookingId);

  await db.runTransaction(async (transaction) => {
    const bookingDoc = await transaction.get(bookingRef);

    if (!bookingDoc.exists) {
      throw new HttpsError("not-found", "Booking not found");
    }

    const booking = bookingDoc.data() as Booking;

    // Authorization: Must be the customer OR a company admin/member
    // A. Check if the user is the customer who made the booking
    let isAuthorized = booking.userId === userId;

    if (!isAuthorized) {
      // B. If not the customer, check if they are a company member
      try {
        const userRef = db.collection("users").doc(userId);
        const userDoc = await transaction.get(userRef);
        
        if (userDoc.exists) {
          const userData = userDoc.data();
          // Check if they are a superadmin
          if (userData?.role === 'super_admin') {
            isAuthorized = true;
          }
          // Check if they belong to the company that owns the booking
          else if (
             userData?.primaryCompanyId === booking.companyId ||
             (Array.isArray(userData?.companyIds) && userData.companyIds.includes(booking.companyId))
          ) {
            isAuthorized = true;
          }
        } else {
           // Fallback to Firebase auth claims (if we can read them, but since we only have userId here)
           const userRecord = await getAuth().getUser(userId);
           if (
              userRecord.customClaims?.role === 'super_admin' ||
              userRecord.customClaims?.companyId === booking.companyId ||
              (Array.isArray(userRecord.customClaims?.companyIds) && userRecord.customClaims.companyIds.includes(booking.companyId))
           ) {
              isAuthorized = true;
           }
        }
      } catch (err) {
        // Ignore auth fetch errors, authorization fails closed.
        console.warn("Failed to check extended permissions for user: ", userId);
      }
    }

    if (!isAuthorized) {
      throw new HttpsError("permission-denied", "You don't have permission to cancel this booking");
    }

    if (booking.status === "cancelled") {
       throw new HttpsError("failed-precondition", "Booking is already cancelled");
    }
    
    if (booking.status === "completed") {
       throw new HttpsError("failed-precondition", "Cannot cancel a completed booking");
    }

    // --- Cancellation Policy Enforcement (pure) ---
    const isCustomer = booking.userId === userId;
    if (isCustomer) {
      try {
        const storeRef = db.doc(`companies/${booking.companyId}/stores/${booking.storeId}`);
        const storeDoc = await transaction.get(storeRef);
        const storeData = storeDoc.exists ? storeDoc.data() : {};
        const policy = BookingPolicySchema.parse(storeData?.bookingPolicy ?? {});

        checkCancellationPolicy({
          bookingStartTime: booking.startTime,
          policy,
        });
      } catch (err) {
        if (err instanceof HttpsError) throw err;
        console.warn("Failed to fetch store for policy check, allowing cancellation:", err);
      }
    }

    // Update the booking status
    const updates: Partial<Booking> = {
      status: "cancelled",
      updatedAt: new Date() as unknown as Date,
    };
    
    if (reason) {
       updates.cancellationReason = reason;
    }

    transaction.update(bookingRef, updates);
  });

  return { success: true };
};

File: DittoDatto-old/services/mercury-engine-old/src/mercury_core/calculators/booking.py
"""MercuryEngine — Booking Receipt & Cancellation Policy.

Replaces:
    packages/mercury-engine/src/core/bookings/booking.ts → buildBookingReceipt
    packages/mercury-engine/src/core/bookings/booking.ts → checkCancellationPolicy

Pure functions — construct a booking snapshot and enforce cancellation rules.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta

from mercury_core.errors import PolicyViolationError
from mercury_core.models.hold import Hold
from mercury_core.models.service import Service
from mercury_core.time import minutes_to_time, parse_time

# ============================================================================
# Booking Receipt
# ============================================================================


@dataclass
class BookingReceiptInput:
    """Input for building a booking receipt from a hold + services."""

    hold: Hold
    services: list[Service]
    user_name: str = ""
    user_email: str = ""
    store_name: str = ""
    payment_id: str | None = None
    notes: str | None = None


@dataclass
class BookingReceiptItem:
    """Single service within the receipt."""

    service_id: str
    title: str
    price: float
    duration: int
    staff_id: str | None = None


@dataclass
class BookingReceipt:
    """Immutable booking receipt (fiscal snapshot)."""

    id: str  # payment_id or generated
    user_id: str
    establishment: str
    service_id: str  # primary service
    status: str = "confirmed"
    start_time: str = ""
    end_time: str = ""
    duration: int = 0
    price_at_time_of_booking: float = 0.0
    service_title: str = ""
    user_name: str = ""
    user_email: str = ""
    store_name: str = ""
    staff_id: str | None = None
    notes: str | None = None
    items: list[BookingReceiptItem] = field(default_factory=list)


def build_booking_receipt(input_data: BookingReceiptInput) -> BookingReceipt:
    """Build an immutable booking receipt from a hold and service catalog.

    This is the snapshot pattern — prices, titles, and durations are frozen
    at the time of booking (Norwegian law, fiscal compliance).

    Args:
        input_data: Hold + services + user info.

    Returns:
        BookingReceipt with all computed fields.
    """
    hold = input_data.hold
    services = input_data.services

    # Resolve service IDs — V1 has serviceIds array, V2 Hold has services list
    service_ids = hold.services if hold.services else []

    # Legacy fallback — if no services list, might have a single service ref
    # (V1 had `serviceId` as well as `serviceIds`)
    if not service_ids:
        # In V2 we don't have a separate serviceId field, but the hold may
        # have been constructed with just one service
        service_ids = []

    # Match services to the hold
    if service_ids:
        matched_services = []
        for sid in service_ids:
            svc = next((s for s in services if s.id == sid), None)
            if svc:
                matched_services.append(svc)
    else:
        # Legacy: use all provided services
        matched_services = list(services)

    # Aggregate
    total_duration = sum(s.duration for s in matched_services)
    total_price = sum(s.price for s in matched_services)
    service_title = ", ".join(s.title for s in matched_services)
    primary_service_id = matched_services[0].id if matched_services else ""

    # Time computation
    slot_start = parse_time(hold.slot_time)
    slot_end = slot_start + total_duration
    start_time = f"{hold.date}T{hold.slot_time}:00"
    end_time = f"{hold.date}T{minutes_to_time(slot_end)}:00"

    # Build items
    items = [
        BookingReceiptItem(
            service_id=s.id or "",
            title=s.title,
            price=s.price,
            duration=s.duration,
            staff_id=hold.staff,
        )
        for s in matched_services
    ]

    # Trim notes
    notes = input_data.notes.strip() if input_data.notes else None
    if notes == "":
        notes = None

    return BookingReceipt(
        id=input_data.payment_id or hold.id or "",
        user_id=hold.user_id,
        establishment=hold.establishment,
        service_id=primary_service_id or "",
        status="confirmed",
        start_time=start_time,
        end_time=end_time,
        duration=total_duration,
        price_at_time_of_booking=total_price,
        service_title=service_title,
        user_name=input_data.user_name,
        user_email=input_data.user_email,
        store_name=input_data.store_name,
        staff_id=hold.staff,
        notes=notes,
        items=items,
    )


# ============================================================================
# Cancellation Policy
# ============================================================================


@dataclass
class CancellationPolicyInput:
    """Policy check input for the V1 contract."""

    client_cancel_enabled: bool = True
    min_cancel_notice_hours: int = 0


def check_cancellation_policy(
    booking_start_time: str | datetime,
    policy: CancellationPolicyInput,
    now: datetime | None = None,
) -> None:
    """Check if a cancellation is allowed under the given policy.

    Args:
        booking_start_time: When the booking starts (ISO string or datetime).
        policy: The cancellation policy to enforce.
        now: Current time (defaults to datetime.now()).

    Raises:
        PolicyViolationError: If cancellation is not allowed.
    """
    if not policy.client_cancel_enabled:
        raise PolicyViolationError("This establishment does not allow customer cancellations")

    if policy.min_cancel_notice_hours <= 0:
        return  # No notice requirement

    # Parse booking start
    if isinstance(booking_start_time, str):
        booking_start = datetime.fromisoformat(booking_start_time)
    else:
        booking_start = booking_start_time

    current_time = now or datetime.now()
    deadline = booking_start - timedelta(hours=policy.min_cancel_notice_hours)

    if current_time > deadline:
        raise PolicyViolationError(
            "Cancellation deadline has passed",
            details={
                "deadline": deadline.isoformat(),
                "notice_hours": policy.min_cancel_notice_hours,
            },
        )
